src/firecracker/src/api_server_adapter.rs

```diff
@@ -161,29 +161,30 @@ pub(crate) fn run_with_api(
         .remove("api")
         .expect("Missing seccomp filter for API thread.");
 
+    let server = match HttpServer::new(&api_bind_path) {
+        Ok(s) => s,
+        Err(ServerError::IOError(inner)) if inner.kind() == std::io::ErrorKind::AddrInUse => {
+            let sock_path = api_bind_path.display().to_string();
+            return Err(ApiServerError::FailedToBindSocket(sock_path));
+        }
+        Err(err) => {
+            return Err(ApiServerError::FailedToBindAndRunHttpServer(err));
+        }
+    };
+
     // Start the separate API thread.
     let api_thread = thread::Builder::new()
         .name("fc_api".to_owned())
         .spawn(move || {
-            match ApiServer::new(to_vmm, from_vmm, to_vmm_event_fd).bind_and_run(
-                &api_bind_path,
+            ApiServer::new(to_vmm, from_vmm, to_vmm_event_fd).run(
+                server,
                 process_time_reporter,
                 &api_seccomp_filter,
                 api_payload_limit,
                 socket_ready_sender,
-            ) {
-                Ok(_) => Ok(()),
-                Err(api_server::Error::ServerCreation(ServerError::IOError(inner)))
-                    if inner.kind() == std::io::ErrorKind::AddrInUse =>
-                {
-                    let sock_path = api_bind_path.display().to_string();
-                    Err(ApiServerError::FailedToBindSocket(sock_path))
-                }
-                Err(api_server::Error::ServerCreation(err)) => {
-                    Err(ApiServerError::FailedToBindAndRunHttpServer(err))
-                }
-            }
-        });
+            );
+        })
+        .expect("API thread spawn failed.");
```

Comment:
Nit: It would be better to also have an error variant for this instead of a possible panic here.